brat/Brat/Checker/Helpers.hs

```diff
@@ -107,27 +109,22 @@ pullPortsSig :: Show ty
              -> Checking [(PortName, ty)]
 pullPortsSig = pullPorts id showSig
 
-pullPorts :: forall a ty. Show ty
-          => (a -> PortName) -- A way to get a port name for each element
-          -> ([(a, ty)] -> String) -- A way to print the list
-          -> [PortName] -- Things to pull to the front
-          -> [(a, ty)]  -- The list to rearrange
-          -> Checking [(a, ty)]
-pullPorts _ _ [] types = pure types
-pullPorts toPort showFn (p:ports) types = do
-  (x, types) <- pull1Port p types
-  (x:) <$> pullPorts toPort showFn ports types
+pullPorts :: forall a ty. (a -> PortName) -- A way to get a port name for each element
+                       -> ([(a, ty)] -> String) -- A way to print the list
+                       -> [PortName] -- Things to pull to the front
+                       -> [(a, ty)]  -- The list to rearrange
+                       -> Checking [(a, ty)]
+pullPorts toPort showFn to_pull types = do -- in Checking
+  (pulled, remaining) <- runStateT (mapM pull1Port to_pull) types
+  pure $ pulled ++ remaining
  where
-  pull1Port :: PortName
-            -> [(a, ty)]
-            -> Checking ((a, ty), [(a, ty)])
-  pull1Port p [] = fail $ "Port not found: " ++ p ++ " in " ++ showFn types
-  pull1Port p (x@(a,_):xs)
-   | p == toPort a
-   = if p `elem` (toPort . fst <$> xs)
-     then err (AmbiguousPortPull p (showFn (x:xs)))
-     else pure (x, xs)
-   | otherwise = second (x:) <$> pull1Port p xs
+  pull1Port :: PortName -> StateT [(a, ty)] Checking (a, ty)
+  pull1Port p = do
+    available <- get
+    case partition ((== p) . toPort . fst) available of
+      ([], _) -> lift $ err $ BadPortPull $ "Port not found: " ++ p ++ " in " ++ showFn available
+      ([found], remaining) -> put remaining >> pure found
+      (_, _) -> lift $ err $ AmbiguousPortPull p (showFn available)
```

Comment:
Not quite a refactor (2/2): we used to pass to `showFn` here only the suffix of `available` beginning with the first match